plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx

```diff
@@ -22,51 +26,60 @@ import ToolbarControls from './toolbar-controls';
 const ProductCollectionContent = (
 	props: ProductCollectionEditComponentProps
 ) => {
-	const { attributes, setAttributes } = props;
-	const { queryId } = attributes;
+	const { clientId, attributes, setAttributes } = props;
 
 	const blockProps = useBlockProps();
 	const innerBlocksProps = useInnerBlocksProps( blockProps, {
 		template: INNER_BLOCKS_TEMPLATE,
 	} );
 
 	const instanceId = useInstanceId( ProductCollectionContent );
-
-	// We need this for multi-query block pagination.
-	// Query parameters for each block are scoped to their ID.
-	useEffect( () => {
-		if ( ! Number.isFinite( queryId ) ) {
-			setAttributes( { queryId: Number( instanceId ) } );
-		}
-	}, [ queryId, instanceId, setAttributes ] );
-
-	/**
-	 * We need to set a unique ID for each instance of this block.
-	 * This will help us uniquely identify each block.
-	 */
-	useEffect( () => {
-		if ( ! attributes?.id ) {
-			setAttributes( { id: uuidV4() } );
-		}
-	}, [ attributes, setAttributes ] );
+	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+	// @ts-ignore These selectors aren't getting their types loaded for some reason.
+	const { getBlock, getBlockParents } = useSelect( blockEditorStore );
 
 	/**
 	 * Because of issue https://github.com/WordPress/gutenberg/issues/7342,
 	 * We are using this workaround to set default attributes.
 	 */
 	useEffect( () => {
+		// In order to properly support pagination this block has a queryId attribute that
+		// is initialized to a unique value when the block is first added to the editor.
+		// We use the `instanceId` for this purpose. It is stable across saves as long
+		// as the order of instances of these blocks in the editor does not change.
+		// The block will be re-indexed in that case, however, this won't cause
+		// any problems since the queryid only has to be stable across client
+		// renders.
+		let queryId = instanceId as number;
+
+		// We DO need to be careful to not change the queryId if the block is part of a
+		// sync pattern. When multiple instances of the same pattern are placed on a
+		// page, updates to one will cause the others to be re-inserted. This will
+		// increment the ID again and trigger a re-insertion of another instance
+		// that updates the ID again, and so on until the browser hangs.
+		const blockParents = getBlockParents( clientId );
+		for ( const parentClientId of blockParents ) {
+			const parent = getBlock( parentClientId );
+			if ( parent && parent.name === 'core/block' ) {
+				queryId = attributes.queryId;
+				break;
+			}
+		}
+
 		setAttributes( {
 			...DEFAULT_ATTRIBUTES,
 			query: {
 				...( DEFAULT_ATTRIBUTES.query as ProductCollectionQuery ),
 				inherit: getDefaultValueOfInheritQueryFromTemplate(),
 			},
 			...( attributes as Partial< ProductCollectionAttributes > ),
+			...{
+				queryId,
+			},
```

Comment:
That can be simplified:

```suggestion
			queryId,
```

Reply:
Oh! You're right! This is some crust from splitting the `id` removal into a separate PR.